pandaserver/dataservice/closer.py

```diff
@@ -0,0 +1,265 @@
+        if self.job.destinationSE == "local" and self.job.prodSourceLabel in ["user", "panda"]:
+            # close non-Rucio destinationDBlock immediately
+            final_status = "closed"
+        elif self.job.lockedby == "jedi" and closer_utils.is_top_level_dataset(destination_data_block):
+            # set it closed in order not to trigger DDM cleanup. It will be closed by JEDI
+            final_status = "closed"
+        elif self.job.produceUnMerge():
+            final_status = "doing"
+        else:
+            # set status to 'tobeclosed' to trigger Rucio closing
+            final_status = "tobeclosed"
+        return final_status
+
+    def perform_vo_actions(self, final_status_dataset: list) -> None:
+        """
+        Perform special actions for vo.
+
+        Args:
+            final_status_dataset (list): The final status dataset.
+        """
+        closer_plugin_class = panda_config.getPlugin("closer_plugins", self.job.VO)
+        if closer_plugin_class is None and self.job.VO == "atlas":
+            # use ATLAS plugin for ATLAS
+            from pandaserver.dataservice.closer_atlas_plugin import (
+                CloserAtlasPlugin,
+            )
+
+            closer_plugin_class = CloserAtlasPlugin
+        if closer_plugin_class is not None:
+            closer_plugin = closer_plugin_class(self.job, final_status_dataset, _logger)
+            closer_plugin.execute()
+
+    def start_activator(self, dataset):
+        """
+        Start the activator
+
+        Args:
+            dataset: Dataset.
+            final_status (str): Final status.
+        """
+        # start Activator
+        if closer_utils.is_not_sub_dataset(dataset.name):
+            if self.job.prodSourceLabel == "panda" and self.job.processingType in ["merge", "unmerge"]:
+                # don't trigger Activator for merge jobs
+                pass
+            else:
+                if self.job.jobStatus == "finished":
+                    activator_thread = Activator(self.task_buffer, dataset)
+                    activator_thread.start()
+                    activator_thread.join()
+        else:
+            # unset flag since another thread already updated
+            # flag_complete = False
+            pass
+
+    # main
+    def run(self):
+        """
+        Main method to run the Closer class. It processes each destination dispatch block,
+        updates the dataset status and finalizes pending jobs if necessary.
+        """
+        try:
+            tmp_log = LogWrapper(_logger,
+                                 f"run-{datetime.datetime.now(datetime.timezone.utc).replace(tzinfo=None).isoformat('/')}-{self.panda_id}")
+            tmp_log.debug(f"Start with job status: {self.job.jobStatus}")
+            flag_complete = True
+            final_status_dataset = []
+
+            for destination_data_block in self.destination_data_blocks:
+                dataset_list = []
```

Comment:
The usage of the dataset_list is a bit funny. It gets added one dataset, sorted, and then reset in the next iteration?